tracing/span.go

```diff
@@ -320,81 +425,116 @@ func (s *Span) CreateLocalChildForContext(ctx context.Context, name string, comp
+		baseSpan: baseSpan{
+			trace: newTrace(s.trace.tracer, name),
+		},
+	}
+
+	child.trace.parentID = s.trace.spanID
+	child.trace.traceID = s.trace.traceID
+	child.trace.sampled = s.trace.sampled
+	child.trace.flags = s.trace.flags
+	return child
 }
 
-// CreateClientChildForContext creates a client child-span with given name and a
-// context that can be used by the client you are creating the span for.
+// CreateLocalChild creates a local child-span with given name and component.
 //
-// A client child-span should be used to make requests to other upstream
-// servers.
+// component is optional local component name and could be empty string.
 //
 // Timestamps, counters, and tags won't be inherited.
 // Parent id will be inherited from the span id,
 // and span id will be randomly generated.
 // Trace id, sampled, and flags will be copied over.
-func (s *Span) CreateClientChildForContext(ctx context.Context, name string) (context.Context, *Span) {
-	child := s.CreateClientChild(name)
-	return context.WithValue(ctx, childSpanKey, child), child
+func (s *baseSpan) CreateLocalChild(name, component string) *LocalSpan {
+	child := newLocalSpan(s.trace.tracer, name, component)
+
+	child.trace.parentID = s.trace.spanID
+	child.trace.traceID = s.trace.traceID
+	child.trace.sampled = s.trace.sampled
+	child.trace.flags = s.trace.flags
+	return child
 }
 
-func (s *Span) createChild(name string, spanType SpanType, component string) *Span {
-	span := newSpan(s.tracer, name, spanType)
-	span.spanType = spanType
-	span.component = component
+// ServerSpan defines a tracing span over an entire request to a server.
+type ServerSpan struct {
+	baseSpan
+}
 
-	span.parentID = s.spanID
-	span.traceID = s.traceID
-	span.sampled = s.sampled
-	span.flags = s.flags
+// SpanType returns the SpanType for a Span.
+func (s *ServerSpan) SpanType() SpanType {
+	return SpanTypeServer
+}
 
-	for _, hook := range s.hooks {
-		if err := hook.OnCreateChild(span); err != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnCreateChild hook error: " + err.Error())
-		}
+// Stop ends the span
+func (s *ServerSpan) Stop(err error) {
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		s.SetTag(ZipkinBinaryAnnotationKeyTimeOut, true)
 	}
-	span.startSpan()
-	return span
 }
 
-// ChildAndThriftContext creates both a client child span and a context can be
-// used by the thrift client code.
-//
-// A thrift client call would look like:
-//
-//     clientCtx, span := parentSpan.ChildAndThriftContext(ctx, "myCall")
-//     result, err := client.MyCall(clientCtx, arg1, arg2)
-//     span.End(ctx, err)
-func (s *Span) ChildAndThriftContext(ctx context.Context, name string) (context.Context, *Span) {
-	ctx, span := s.CreateClientChildForContext(ctx, name)
-	ctx = CreateThriftContextFromSpan(ctx, span)
-	return ctx, span
+// SetServerSpan sets the Span as the ServerSpan on the given context
+func (s *ServerSpan) SetServerSpan(ctx context.Context) context.Context {
```

Comment:
nit: Since this is a function of type `ServerSpan`, the name `SetServerSpan` feels redundant, maybe call it `InjectContext` or something?